### PR TITLE
Add support for Amazon CloudWatch Logs PutDestination/PutDestinationPolicy

### PR DESCRIPTION
**Pull request by bodgit**
This implements the missing resources outlined in #9930. I've added the following resources:

* `aws_cloudwatch_log_destination`
* `aws_cloudwatch_log_destination_policy`

This is an example minimal Terraform manifest that mimics the [Amazon documentation](http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CrossAccountSubscriptions.html):

```
variable "aws_access_key" {}

variable "aws_secret_key" {}

variable "aws_region" {
  default = "eu-west-2"
}

provider "aws" {
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
  region     = "${var.aws_region}"
}

resource "aws_kinesis_stream" "test" {
  name        = "RootAccess"
  shard_count = 1
}

data "aws_iam_policy_document" "role" {
  statement {
    effect = "Allow"

    principals = {
      type = "Service"

      identifiers = [
        "logs.${var.aws_region}.amazonaws.com",
      ]
    }

    actions = [
      "sts:AssumeRole",
    ]
  }
}

resource "aws_iam_role" "test" {
  name               = "CWLtoKinesisRole"
  assume_role_policy = "${data.aws_iam_policy_document.role.json}"
}

data "aws_iam_policy_document" "policy" {
  statement {
    effect = "Allow"

    actions = [
      "kinesis:PutRecord",
    ]

    resources = [
      "${aws_kinesis_stream.test.arn}",
    ]
  }

  statement {
    effect = "Allow"

    actions = [
      "iam:PassRole",
    ]

    resources = [
      "${aws_iam_role.test.arn}",
    ]
  }
}

resource "aws_iam_role_policy" "test" {
  name   = "Permissions-Policy-For-CWL"
  role   = "${aws_iam_role.test.id}"
  policy = "${data.aws_iam_policy_document.policy.json}"
}

resource "aws_cloudwatch_log_destination" "test" {
  name       = "testDestination"
  target_arn = "${aws_kinesis_stream.test.arn}"
  role_arn   = "${aws_iam_role.test.arn}"
}

data "aws_iam_policy_document" "access" {
  statement {
    effect = "Allow"

    principals = {
      type = "AWS"

      identifiers = [
        "123456789012",
      ]
    }

    actions = [
      "logs:PutSubscriptionFilter",
    ]

    resources = [
      "${aws_cloudwatch_log_destination.test.arn}",
    ]
  }
}

resource "aws_cloudwatch_log_destination_policy" "test" {
  destination_name = "testDestination"
  access_policy    = "${data.aws_iam_policy_document.access.json}"
}

output "destination" {
  value = "${aws_cloudwatch_log_destination.test.arn}"
}
```

It's still missing tests and documentation however it seems to be working in the limited testing I've given it and I was wondering if someone could give the code a review to make sure I'm not doing anything stupid?

I'm a little unsure about the `aws_cloudwatch_log_destination_policy` resource as the API only offers a dedicated `PutDestinationPolicy` action, so to read the policy back in I use `DescribeDestinations` and pick out the policy from that. Also there doesn't seem to be a way to delete a policy, merely overwrite it so the delete is currently a no-op as I can't seem to overwrite it with anything other than another valid policy, I'm not sure what to do here.

### CODE REVIEW

**Comment by bodgit** [comment]
One problem I keep getting and I don't think it's specific to do with my new resources is that every time I try and apply the above example manifest from clean it always fails trying to create the destination bound to the Kinesis stream. If I add an additional `aws_cloudwatch_log_subscription_filter` resource like so:

```
resource "aws_cloudwatch_log_group" "test" {
  name = "CloudTrail"
}

resource "aws_cloudwatch_log_subscription_filter" "test" {
  name = "RootAccess"
  role_arn = "${aws_iam_role.test.arn}"
  log_group_name = "${aws_cloudwatch_log_group.test.name}"
  filter_pattern = "{$.userIdentity.type = Root}"
  destination_arn = "${aws_kinesis_stream.test.arn}"
}
```
with it bound to the same Kinesis stream then I get the following error on the first run:
```
* aws_cloudwatch_log_subscription_filter.test: 1 error(s) occurred:

* aws_cloudwatch_log_subscription_filter.test: InvalidParameterException: Could not deliver test message to specified Kinesis stream. Check if the given kinesis steam is in ACTIVE state.
	status code: 400, request id: 2a9ac3d9-f2ef-11e6-902f-5975afceb44f
```

The Kinesis stream is shown as ACTIVE and if I retry the apply immediately again then the remaining resources are created without problem.

This will make testing difficult as it's never going to work cleanly in one run and I don't think I can bind a destination to anything other than a Kinesis stream.

**Comment by bodgit** [comment]
I've improved the error reporting so I now get the same sort of error message as for a subscription filter. Full run looks like this:

```
data.aws_iam_policy_document.role: Refreshing state...
aws_cloudwatch_log_group.test: Creating...
  arn:               "" => "<computed>"
  name:              "" => "CloudTrail"
  retention_in_days: "" => "0"
aws_kinesis_stream.test: Creating...
  arn:              "" => "<computed>"
  name:             "" => "RootAccess"
  retention_period: "" => "24"
  shard_count:      "" => "1"
aws_iam_role.test: Creating...
  arn:                "" => "<computed>"
  assume_role_policy: "" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"\",\n      \"Effect\": \"Allow\",\n      \"Action\": \"sts:AssumeRole\",\n      \"Principal\": {\n        \"Service\": \"logs.eu-west-2.amazonaws.com\"\n      }\n    }\n  ]\n}"
  create_date:        "" => "<computed>"
  name:               "" => "CWLtoKinesisRole"
  path:               "" => "/"
  unique_id:          "" => "<computed>"
aws_cloudwatch_log_group.test: Creation complete
aws_iam_role.test: Creation complete
aws_kinesis_stream.test: Still creating... (10s elapsed)
aws_kinesis_stream.test: Still creating... (20s elapsed)
aws_kinesis_stream.test: Still creating... (30s elapsed)
aws_kinesis_stream.test: Still creating... (40s elapsed)
aws_kinesis_stream.test: Still creating... (50s elapsed)
aws_kinesis_stream.test: Creation complete
aws_cloudwatch_log_destination.test: Creating...
  arn:        "" => "<computed>"
  name:       "" => "testDestination"
  role_arn:   "" => "arn:aws:iam::123456789012:role/CWLtoKinesisRole"
  target_arn: "" => "arn:aws:kinesis:eu-west-2:123456789012:stream/RootAccess"
aws_cloudwatch_log_subscription_filter.test: Creating...
  destination_arn: "" => "arn:aws:kinesis:eu-west-2:123456789012:stream/RootAccess"
  filter_pattern:  "" => "{$.userIdentity.type = Root}"
  log_group_name:  "" => "CloudTrail"
  name:            "" => "RootAccess"
  role_arn:        "" => "arn:aws:iam::123456789012:role/CWLtoKinesisRole"
data.aws_iam_policy_document.policy: Refreshing state...
aws_iam_role_policy.test: Creating...
  name:   "" => "Permissions-Policy-For-CWL"
  policy: "" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"\",\n      \"Effect\": \"Allow\",\n      \"Action\": \"kinesis:PutRecord\",\n      \"Resource\": \"arn:aws:kinesis:eu-west-2:123456789012:stream/RootAccess\"\n    },\n    {\n      \"Sid\": \"\",\n      \"Effect\": \"Allow\",\n      \"Action\": \"iam:PassRole\",\n      \"Resource\": \"arn:aws:iam::123456789012:role/CWLtoKinesisRole\"\n    }\n  ]\n}"
  role:   "" => "CWLtoKinesisRole"
aws_iam_role_policy.test: Creation complete
Error applying plan:

2 error(s) occurred:

* aws_cloudwatch_log_subscription_filter.test: 1 error(s) occurred:

* aws_cloudwatch_log_subscription_filter.test: InvalidParameterException: Could not deliver test message to specified Kinesis stream. Check if the given kinesis steam is in ACTIVE state.
	status code: 400, request id: 8a894d60-f361-11e6-8c45-7b03d96772ce
* aws_cloudwatch_log_destination.test: 1 error(s) occurred:

* aws_cloudwatch_log_destination.test: InvalidParameterException: Could not deliver test message to specified destination. Check if the destination is valid.
	status code: 400, request id: 8a8925c9-f361-11e6-9841-73821809b9e1

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

According to https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_kinesis_stream.go#L84 it looks like the Kinesis creation should block until it is ACTIVE and on my test runs it does sit there for about a minute before continuing with the next resource.

**Comment by bodgit** [comment]
I've found two issues, one I wasn't depending on the IAM role policy before creating the log destination and/or subscription filter so I don't think it would naturally work in the right order. I ended up adding something like this to both resources:
```
resource "aws_cloudwatch_log_destination" "test" {
  name = "testDestination"
  target_arn = "${aws_kinesis_stream.test.arn}"
  role_arn = "${aws_iam_role.test.arn}"
  depends_on = ["aws_iam_role_policy.test"]
}
```
This at least means everything is created in the correct order however there's still some sort of internal API delay on the Amazon side where all dependencies are created but I still can't create a destination or subscription filter without getting the above error and it will persist for a few minutes before it finally succeeds without any changes on my part. To work around this I've handled that specific error message and made the creation code retry (this change has also been applied to the subscription filter resource) and I've also increased the timeout so it will retry for longer until it finally gives up if Amazon won't play ball.

At least now I can try and write some acceptance tests.

**Comment by bodgit** [comment]
I've now added tests:

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCloudwatchLogDestination'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/15 14:04:37 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCloudwatchLogDestination -timeout 120m
=== RUN   TestAccAWSCloudwatchLogDestinationPolicy_importBasic
--- PASS: TestAccAWSCloudwatchLogDestinationPolicy_importBasic (91.53s)
=== RUN   TestAccAWSCloudwatchLogDestination_importBasic
--- PASS: TestAccAWSCloudwatchLogDestination_importBasic (101.80s)
=== RUN   TestAccAWSCloudwatchLogDestinationPolicy_basic
--- PASS: TestAccAWSCloudwatchLogDestinationPolicy_basic (106.13s)
=== RUN   TestAccAWSCloudwatchLogDestination_basic
--- PASS: TestAccAWSCloudwatchLogDestination_basic (103.85s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	403.341s
```

With that it should be good to go, pending any further feedback.

File: builtin/providers/aws/resource_aws_cloudwatch_log_destination.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsCloudWatchLogDestination() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCloudWatchLogDestinationPut,
+		Update: resourceAwsCloudWatchLogDestinationPut,
+
```
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_cloudwatch_log_destination.go]: Remove odd spacing here

File: builtin/providers/aws/resource_aws_cloudwatch_log_destination_test.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+resource "aws_iam_role_policy" "test" {
+  name = "Permissions-Policy-For-CWL"
+  role = "${aws_iam_role.test.id}"
+  policy = "${data.aws_iam_policy_document.policy.json}"
+}
+
+resource "aws_cloudwatch_log_destination" "test" {
+  name = "testDestination"
```
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_cloudwatch_log_destination_test.go]: Mind seeding all these test value names? We run our acceptance tests in parallel on multiple branches sometimes, and conflicting names will cause the tests to fail.

**Comment by bodgit** [comment]
Changes made as requested, tests still pass:
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCloudwatchLogDestination'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/18 18:45:35 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCloudwatchLogDestination -timeout 120m
=== RUN   TestAccAWSCloudwatchLogDestinationPolicy_importBasic
--- PASS: TestAccAWSCloudwatchLogDestinationPolicy_importBasic (96.02s)
=== RUN   TestAccAWSCloudwatchLogDestination_importBasic
--- PASS: TestAccAWSCloudwatchLogDestination_importBasic (91.90s)
=== RUN   TestAccAWSCloudwatchLogDestinationPolicy_basic
--- PASS: TestAccAWSCloudwatchLogDestinationPolicy_basic (96.32s)
=== RUN   TestAccAWSCloudwatchLogDestination_basic
--- PASS: TestAccAWSCloudwatchLogDestination_basic (95.27s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	379.545s
```

**Pull request merged**